### DISCUSSION TITLE
Can zustand persist() work within an SSR app?

### DISCUSSION THREAD

**Discussion by piscopancer** [pmndrs/zustand#1797] [Q&A]
Persist refers to local storage as a browser API so it is unreachable during SSR. The page is loaded during SSR with default values or fetched values (we all know what) and only then does Zustand does its work checking for data in local storage and updating the state thanks to persist() function. I have not used Zustand with NextJS. I wonder if values are going to flick and shake🤔

Why am I asking? I am currently working with Qwik framework and Zustand is incompatible. Qwik uses SSR and I am using loading data from local storage to the state. However SSR loads in the 'default' values of store and only after SSR qwik requires ~0.1s to run all javascript and only then local storage can be checked for values. So the flicker or flashing of values takes ~0.1s and that's annoying. There is not flicker in zustand.

**Comment by IMythx** [comment]
it flickers in zustand too...it still take sometime to read the data from the localstorage

**Reply by dai-shi** [reply]
Yeah, and it's not only Zustand. localStorage is only available on the client.

**Reply by IMythx** [reply]
yeah its not zustand specific thing

the best solution i could found is to store the data in the cookies too, read it in a server component and then pass it down to the client component
this way the data will be rdy when the component renders without the need to wait for the localstorage to be rdy

**Comment by mikhkosvn** [comment]
not perfect but at least no flickering (have some delay before text is shown)

used onRehydrateStorage and showed text only after rehidration 


```
import { create } from 'zustand';
import { persist } from 'zustand/middleware';


interface DropdownTestStore {
  selectedTestType: 'automated-test' | 'manual-test';
  setTestType: (type: 'automated-test' | 'manual-test') => void;
  isRehydrated: boolean;
  setRehydrated: (rehydrated: boolean) => void;
}


export const useTestTypeDropdown = create<DropdownTestStore>()(
  persist(
    (set) => ({
      selectedTestType: 'automated-test',
      setTestType: (type) => set({ selectedTestType: type }),
      isRehydrated: false,
      setRehydrated: (rehydrated) => set({ isRehydrated: rehydrated }),
    }),
    {
      name: 'test-type-storage',
      onRehydrateStorage: () => (state) => {
        state?.setRehydrated(true); // set rehydrated to true when persistence is complete
      },
    },
  ),
);

```
